.github/workflows/windows-workflow.yml

```
@@ -94,24 +99,59 @@ jobs:
         if: steps.cache-submodules.outputs.cache-hit != 'true'
         run: git submodule update --init --recursive -j $env:NUMBER_OF_PROCESSORS
 
+      - name: Setup Buildcache
+        uses: mikehardy/buildcache-action@v1.2.1
+        with:
+          cache_key: ${{ matrix.os }} ${{ matrix.platform }} ${{ matrix.configuration }}
+        if: ${{ matrix.configuration == 'CMake' }} # TODO: buildcache on VS
+
       - name: Verify VS Project Files
         run: .github\workflows\scripts\windows\validate-vs-filters.ps1
+        if: ${{ matrix.configuration != 'CMake' }}
 
       - name: Setup msbuild
         uses: microsoft/setup-msbuild@v1
+        if: ${{ matrix.configuration != 'CMake' }}
+
+      - name: Generate CMake
+        id: cmake
+        shell: cmd
+        run: |
+          if "${{ github.event.inputs.retainDebugArtifacts }}"=="true" (SET type=RelWithDebInfo) else (SET type=Release)
+          if "${{ matrix.platform }}"=="Win32" (SET vcvars=vcvarsamd64_x86.bat) else (SET vcvars=vcvars64.bat)
+          call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\%vcvars%"
+          echo ::set-output name=buildtype::%type%
+          echo ::set-output name=vcvars::%vcvars%
+          mkdir build
+          cd build
+          cmake .. -DCMAKE_BUILD_TYPE=%type% -G Ninja -DCMAKE_C_COMPILER_LAUNCHER=..\buildcache\bin\buildcache.exe -DCMAKE_CXX_COMPILER_LAUNCHER=..\buildcache\bin\buildcache.exe -DCMAKE_DISABLE_PRECOMPILE_HEADERS=ON
```

Not sure if this will work or not.
```suggestion
          cmake -B build -DCMAKE_BUILD_TYPE=%type% -G Ninja -DCMAKE_C_COMPILER_LAUNCHER=buildcache\bin\buildcache.exe -DCMAKE_CXX_COMPILER_LAUNCHER=buildcache\bin\buildcache.exe -DCMAKE_DISABLE_PRECOMPILE_HEADERS=ON
```